### PR TITLE
Remove any use of Regex in the Version.Parser module

### PR DESCRIPTION
**Pull request by whatyouhide**

### CODE REVIEW

File: lib/elixir/lib/version.ex [hunk 1]
```diff
@@ -392,67 +392,93 @@ defmodule Version do
...
-          patch = patch && String.to_integer(patch)
-
-          case parse_pre(pre) do
-            {:ok, pre} ->
-              {:ok, {major, minor, patch, pre}}
-            :error ->
-              :error
-          end
-        end
+      destructure [version_with_pre, build_metadata], String.split(string, "+", parts: 2)
```
Comment (josevalim) [inline lib/elixir/lib/version.ex]: Why not?

    [version_with_pre | build_metadata] = String.split(string, "+")

Comment (whatyouhide) [inline lib/elixir/lib/version.ex]: Only one `+` is allowed. I don't see the advantage of doing that :|

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

File: lib/elixir/lib/version.ex [hunk 2]
```diff
@@ -392,67 +392,92 @@ defmodule Version do
...
-          case parse_pre(pre) do
-            {:ok, pre} ->
-              {:ok, {major, minor, patch, pre}}
-            :error ->
-              :error
-          end
-        end
+      destructure [version_with_pre, build], String.split(string, "+", parts: 2)
+      destructure [version, pre], String.split(version_with_pre, "-", parts: 2)
+      destructure [major, minor, patch], String.split(version, ".")
```
Comment (ericmj) [inline lib/elixir/lib/version.ex]: Do we have a test case for "1.2.3.4"? I don't see how we catch this.

Comment (josevalim) [inline lib/elixir/lib/version.ex]: Good catch. Fixed.

**Pull request merged**